### PR TITLE
Tweak stream dependency processor

### PR DESCRIPTION
**Pull request by dsyer**
If user chooses anything stream-related apply the same logic, mainly
to pick up the binder if "kafka" or "rabbitmq" are selected.

### CODE REVIEW

File: initializr-service/src/main/java/io/spring/initializr/service/InitializrService.java [hunk 1]
```diff
@@ -46,7 +46,6 @@ public static void main(String[] args) {
 	}
 
 	@Bean
-	@SuppressWarnings("deprecation")
```
Comment (snicoll) [inline initializr-service/src/main/java/io/spring/initializr/service/InitializrService.java:49]: That  change is unrelated (and I am not sure I'd merge that anyway)

File: initializr-service/src/main/resources/application.yml [hunk 2]
```diff
@@ -719,11 +722,14 @@ initializr:
         - name: Kafka
           id: kafka
           weight: 100
-          description: Kafka messaging support using Spring Kafka
+          description: Kafka messaging support using Sring Kafka
```
Comment (snicoll) [inline initializr-service/src/main/resources/application.yml]: ?

File: initializr-service/src/main/java/io/spring/initializr/service/extension/SpringCloudStreamRequestPostProcessor.java [hunk 3]
```diff
@@ -33,38 +33,53 @@
 
 	private final Dependency springCloudStreamKafka;
 
-	private final Dependency springCloudStreamReactive;
+	private final Dependency springCloudStreamRabbit;
+
+	private final Dependency streamTest;
```
Comment (snicoll) [inline initializr-service/src/main/java/io/spring/initializr/service/extension/SpringCloudStreamRequestPostProcessor.java]: I am not against it but I'd like this issue to be focused to the task at hand.

File: initializr-service/src/main/resources/application.yml [hunk 4]
```diff
@@ -783,7 +789,7 @@ initializr:
           artifactId: spring-cloud-starter-oauth2
         - name: Cloud Task
           id: cloud-task
-          description: Task result tracking along with integration with batch and streams
+          description: Task result tracking and integration with Spring Batch
```
Comment (snicoll) [inline initializr-service/src/main/resources/application.yml:792]: Unrelated change?!

File: initializr-service/src/main/resources/application.yml [hunk 5]
```diff
@@ -1011,28 +1017,24 @@ initializr:
...
+          description: Reactive messaging microservices with Spring Cloud Stream
           versionRange: 1.3.0.RELEASE
           groupId: org.springframework.cloud
-          artifactId: spring-cloud-starter-stream-rabbit
-        - name: Cloud Kafka Streams
+          artifactId: spring-cloud-stream-reactive
+        - name: Cloud Stream Kafka Streams
           id: cloud-stream-binder-kafka-streams
-          description: Stream processing microservices with Spring Cloud Stream and Kafka Streams API
+          description: Stream processing microservices with Spring Cloud Stream and Kafka Streams API (not needed for vanilla Spring Cloud Stream with Kafka)
```
Comment (snicoll) [inline initializr-service/src/main/resources/application.yml]: I know the "not needed for vanilla Spring Cloud Stream with Kafka" is supposed to help but it  confuses me. Why isn't that needed? Why do we have to state that?

File: initializr-service/src/main/resources/application.yml [hunk 6]
```diff
@@ -1011,28 +1017,24 @@ initializr:
...
         - name: Cloud Stream
           id: cloud-stream
-          aliases:
-            - cloud-stream-binder-rabbit
-          description: Messaging microservices with Spring Cloud Stream and RabbitMQ
+          description: Messaging microservices with Spring Cloud Stream (requires a binder, e.g. Kafka or RabbitMQ)
           versionRange: 1.3.0.RELEASE
           groupId: org.springframework.cloud
-          artifactId: spring-cloud-starter-stream-rabbit
+          artifactId: spring-cloud-stream
```
Comment (snicoll) [inline initializr-service/src/main/resources/application.yml]: So we're at the core of my question here. What happens if I had `spring-kafka` here? Do I get _exactly_ the same behaviour as adding the previous kafka binder? 

(I know the answer to my own question).

If the concept does not exist in the project, start.spring.io should not pretend it exists. If that's what we want, it needs to be a first class concept in Spring Cloud Stream.

Depending on you are selecting a binder, a non starter becomes a starter. This is convoluted.

File: initializr-service/src/main/resources/application.yml [hunk 7]
```diff
@@ -1011,28 +1017,24 @@ initializr:
       content:
         - name: Cloud Bus
           id: cloud-bus
-          aliases:
-            - cloud-bus-amqp
-          description: A simple control bus defaulting to RabbitMQ
+          description: A simple control bus using Spring Cloud Stream (requires a binder, e.g. Kafka or RabbitMQ)
           groupId: org.springframework.cloud
-          artifactId: spring-cloud-starter-bus-amqp
+          artifactId: spring-cloud-bus
```
Comment (snicoll) [inline initializr-service/src/main/resources/application.yml]: Same comment on this one. This is a "fake entry"

**Review by snicoll — CHANGES_REQUESTED** [review]
I don't mind that  we don't bring a default binder by default but  the  project should support that use case. I  am -1 to change initializr to pretend that it does.

File: initializr-service/src/main/java/io/spring/initializr/service/InitializrService.java (hunk 1 above)
Comment (dsyer) [reply initializr-service/src/main/java/io/spring/initializr/service/InitializrService.java:49]: It was showing as a warning (not needed) in my perfect IDE. It's unrelated, but I don't see any reason not to make this change.

File: initializr-service/src/main/java/io/spring/initializr/service/extension/SpringCloudStreamRequestPostProcessor.java (hunk 3 above)
Comment (dsyer) [reply initializr-service/src/main/java/io/spring/initializr/service/extension/SpringCloudStreamRequestPostProcessor.java]: Not sure. I thought the task at hand was tidying up the stream dependencies.

File: initializr-service/src/main/resources/application.yml (hunk 2 above)
Comment (dsyer) [reply initializr-service/src/main/resources/application.yml]: Copy-paste I guess.

File: initializr-service/src/main/resources/application.yml (hunk 4 above)
Comment (dsyer) [reply initializr-service/src/main/resources/application.yml:792]: Not really. It focuses the search in the web UI better on stream things for stream apps.

File: initializr-service/src/main/resources/application.yml (hunk 7 above)
Comment (dsyer) [reply initializr-service/src/main/resources/application.yml]: It doesn't have to be, Spring Cloud Bus has a jar file.

File: initializr-service/src/main/resources/application.yml (hunk 6 above)
Comment (dsyer) [reply initializr-service/src/main/resources/application.yml]: I thought I was just copying your logic (removing the core jar and adding a starter). I'm quite happy to leave it in (so core + kafka). It's just like the GCP SQL use case (the starters end up not being used by Initializr).

File: initializr-service/src/main/resources/application.yml (hunk 5 above)
Comment (dsyer) [reply initializr-service/src/main/resources/application.yml]: I want to prevent people from choosing this one when they don't know what it is, but they do know they want Streams + Kafka.

**Comment by wilkinsona** [comment]
@snicoll asked me to take a look at this so I've been trying to educate myself on all of the pieces that are involved so I can offer an informed opinion. Doing so has left me more confused than when I started. For example, why does [spring-cloud-starter-stream-rabbit](https://github.com/spring-cloud/spring-cloud-stream-binder-rabbit/blob/master/spring-cloud-starter-stream-rabbit/pom.xml) only have a single dependency? It's far from being a typical starter module.

From what I've seen thus far, Spring Cloud Stream seems to be structured in an atypical manner and we're trying to use Initializr to "fix" that. Where can I learn about the rationale behind that structure and why fixing it in Initializr rather than in SCS is the approach that we want to take?

**Comment by dsyer** [comment]
I expect the starter is technically redundant. Think of it as a "marketing" starter, or a "discoverability" starter (people know how to find dependencies that start with "spring-cloud-starter"). It doesn't even need to be used in Initializr.

I don't think that Spring Cloud Stream is perfect, but this change is not about fixing or hiding anything there. It is rationalizing the checkboxes and making the user choices and search results in the UI simpler and easier to understand.

**Comment by wilkinsona** [comment]
> It is rationalizing the checkboxes and making the user choices and search results in the UI simpler and easier to understand.

Can you please help me to understand why SCS is structured in such a way that custom logic in Initializr is required? I'd also like to know if structuring SCS differently has been considered. If it has, why is changing Initializr the preferred approach?

**Comment by dsyer** [comment]
I don't yet see how structuring SCS differently would solve any problem, but please feel free to make a proposal, even if it's only a straw man. It's probably a bit like Spring Session. I haven't analysed that one in detail but there is a similar feature in Initializr there, so the analogy might be useful even if they are not identical.

You have a core set of features (and autoconfiguration FWIW), and then an optional set of "binders" (middleware choices). Those middleware choices also make sense for vanilla Spring Boot apps, not using Spring Cloud Stream, so the main feature we are adding to the Initializr is that middleware checkbox (Rabbit or Kafka at this point in time, plenty of options for others later). It's an important simplification that users can select Kafka and may or may not select Spring Cloud Stream, and in both cases it makes sense. Essentially it reduces the number of checkboxes, and makes it easier for users to reason about what is going to be generated. It's even more important when you consider that there are other Spring Cloud features built on top of Streams (Bus and Turbine Stream) which, without this new feature, require a checkbox per middleware, and now only require one.

**Comment by snicoll** [comment]
>  It's probably a bit like Spring Session

I guess we all remember that we spent a lot of time discussing about that one and the outcome was a compromise and certainly not what the majority of us would have liked. Just because you were (maybe) in the other camp isn't a reason to use it as a justification for this one.

>  the main feature we are adding to the Initializr is that middleware checkbox

I think the end goal is clear for everyone and I don't see how it answers the question @wilkinsona already asked twice.

**Comment by dsyer** [comment]
If the goal is clear, and the PR delivers that, then I don't see how the questions are relevant (even if I knew what you meant).

**Comment by wilkinsona** [comment]
The questions are relevant (to me at least) as I'm hoping that the answers to them will help me to understand the problem and why the changes that are proposed here are the right way to solve it. In other words, this PR may deliver the goal but I can't tell if the goal needs to be delivered in this way. The current proposal seems less than ideal as it requires custom logic and is repeating some of the problems we had with Spring Session. Right now, a project structure that maps more directly to check boxes (as we used to have with Spring Session) would seem to be preferable for Intializr and users in general.

**Comment by dsyer** [comment]
I don't think a project structure that maps directly onto check boxes is always desirable. This might be a good example of why it is not.

Here is a schematic of the checkboxes if you organize checboxes to correspond directly to jar files:

```
Cloud Streams
[x] Stream Rabbit
[ ] Stream Kafka
...

Circuit Breaker
[ ] Turbine Rabbit
[ ] Turbine Kafka
...

Cloud Messaging
[ ] Bus Rabbit
[ ] Bus Kafka
...
```

and here's what happens if you allow the middleware (binder) to float as a separate choice, as proposed here:

```
Cloud Streams
[x] Stream

Circuit Breaker
[ ] Turbine

Cloud Messaging
[ ] Bus

Middleware
[x] Rabbit
[ ] Kafka
...
```

It's fewer checkboxes (by one, so not a big deal yet), but it is also much easier to understand for users and much easier to pick out what you need. But imagine what happens if you increase the number of available binders from 2 to N, and the number of integrations from 3 to M. With checkbox per jar you have M*N checkboxes (currently 6), and with the middleware selection it is M+N (currently 5). I'd argue that even without the expansion (which is not unlikely), it's a valuable simplification.

It's not impossible to build an app with vanilla streams binding to Rabbit and Bus stream binding to Kafka, but I'd argue that is advanced usage, and unlikely to be covered by Initializr out of the box, so whether as a user I have to choose 3 checkboxes or 2 is kind of irrelevant.

If there are any more questions I didn't answer, please just ask them again (out of band if you like).

**Comment by wilkinsona** [comment]
When a project is structured the way the SCS is currently structured, I agree that mapping checkboxes to jar files isn't desirable. What I am trying to understand is why we appear to be jumping to the conclusion the Initializr is the place to "fix" that structure rather than making a change to the project itself.

Any clever hoops that Initializr jumps through on a user's behalf are hoops that a user will have to jump through themselves if they're not using Initializr to configure their dependencies. Does the M*N argument not also apply to Spring Cloud Stream's starters given the current arrangement? Put another way, would it not be better for everyone if a SCS user could change their pom to add a dependency on `spring-cloud-stream-starter` and then a middleware dependency? That would mean that all SCS users get M+N irrespective of whether or not they're using Initializr.

**Comment by dsyer** [comment]
That would only be the case if it made technical sense to smash all of Spring Cloud Streams (and the things that integrate with it) into a big Frankenjar. I don't think anyone's interests are well served by that. It's too much happening in one place and it has to be up to the projects if they want to keep things separate. The release cadence of Turbine is quite different than Spring Cloud Stream. And the Kafka binder evolves completely indpendently of the Rabbit binder. New binders can be developed as standalone libraries. I don't see any downside in that really. Initializr is really just simplifying a choice that you can make yourself quite easily (essentially adding an extra jar that you didn't explicitly choose, or swapping one for another).

**Comment by wilkinsona** [comment]
Thanks. I think we're getting somewhere now.

>  I don't see any downside in that really

A downside is that, outside of Initializr, users have more dependencies to worry about. The SCS team presumably weighed up the pros and cons of the possible structures and decided that M*N was better than M+N as it gave them more flexibility in terms of release schedule, etc.

> Initializr is really just simplifying a choice that you can make yourself quite easily

Right, and this is where things get really subjective. For me, if the choice is one that can be made quite easily outside of Initializr, then Initializr should just present that choice without adding custom logic to smooth over some of the downsides of the project's structure. I dislike the current situation with Spring Session for the same reason and would prefer not to have that custom logic too.

**Comment by dsyer** [comment]
I'm not sure in practice it really is M*N outside of Initializr; it's more like M+N (choose an integration and then choose a binder). In Initializr we are trying to achieve a harder goal of simplifying the choice for users who are discovering new things by allowing them to do it with one checkbox for simple use cases. Actually we (or I anyway) decided that one checkbox selection covering all of the choices here is impractical. I.e. this would work just as well

```
Cloud Streams
[x] Stream
[x] Stream Rabbit
[ ] Stream Kafka

Circuit Breaker
[ ] Turbine

Cloud Messaging
[ ] Bus
```

But I thought we wanted to offer the "middleware" choice as a valid option for non stream apps. If no-one really cares about that then I guess I don't mind, and we can probably remove the post processor completely (it would only be there as a courtesy to helpfully remove a redundant transitive dependency).

**Comment by joshlong** [comment]
Are we proposing not letting people choose Spring AMQP and Spring Kafka, unless they bring in through Spring Cloud Stream?

**Comment by dsyer** [comment]
No, those choices would still have to be there (and we need to check that this PR doesn't duplicate them or remove them). The original proposal was to use those as a trigger to include a Stream binder, if the user also chose a Stream project, and that would make the binder checkboxes redundant. Andy feels like that is too much magic if I understood correctly, but it still makes sense to me.

**Comment by joshlong** [comment]
I think the magic would be very useful. It would help them make the choice more easily. But, if removing the magic would allow the reactive Spring Cloud Stream checkbox to be merged soon, then I’d prefer that

**Comment by philwebb** [comment]
This is a pretty interesting discussion and I realized that we've never really documented our recommendations around starters and auto-configuration. I've made a start [with a WIKI page on the subject](https://github.com/spring-projects/spring-boot/wiki/Building-On-Spring-Boot).

> The Kafka binder evolves completely indpendently of the Rabbit binder

I do understand the logic here, but I often find it a little confusing as a consumer trying to match versions together. Since it's possible to have a `spring-cloud-stream` release that doesn't have a  related `spring-cloud-stream-binder-kafka` (and visa-versa) I assume the rule should be I always declare both dependencies and the latest patch version? In other words, I shouldn't do this:

```
<dependency>
	<groupId>org.springframework.cloud</groupId>
	<artifactId>spring-cloud-stream-binder-kafka</artifactId>
	<version>2.0.4</version>
</dependency>
```

Because the transitive dependency to `spring-cloud-stream` might not be the latest. I need to do:

```
<dependency>
	<groupId>org.springframework.cloud</groupId>
	<artifactId>spring-cloud-stream</artifactId>
	<version>2.0.4</version>
</dependency>
<dependency>
	<groupId>org.springframework.cloud</groupId>
	<artifactId>spring-cloud-stream-binder-kafka</artifactId>
	<version>2.0.4</version>
</dependency>
```

And v2.0.4 of the Kafka binder might actually be older than 2.0.4 of Stream (i.e. It could pull in Stream 2.0.1).

I assume there's no plans for a BOM or a release train with Spring Cloud Stream?

BTW, I think this pattern is slightly different from Spring Session where they have all supported stores in the [same repo](https://github.com/spring-projects/spring-session) and release them at the same time. Although, they do also have [spring-session-data-mongodb](https://github.com/spring-projects/spring-session-data-mongodb) which is in a different repo for some reason.

I guess if this pattern of a specific starter when combining things is going to be used often we should make some changes to the initializr meta-data. Needing a custom post-processor for each thing feels like something that could bite us later.

**Comment by philwebb** [comment]
Oh wait. I see [there is a BOM](https://github.com/spring-cloud/spring-cloud-stream-starters/blob/master/spring-cloud-stream-dependencies/pom.xml). There's also a [BOM for Spring Session](https://github.com/spring-projects/spring-session-bom).

**Comment by snicoll** [comment]
Interesting this one is stil opened. We've moved on and merged a first iteration. We're working on the second in https://github.com/spring-io/initializr/issues/642
